_posts/2023-01-14-trivially-relocatable.md

```diff
@@ -0,0 +1,656 @@
+
+#### Why do we care about data size?
+
+We care about "data size" because
+when the "data size" of a type differs from its "size", it becomes a
+[potentially overlapping type](
+https://quuxplusone.github.io/blog/2018/07/13/trivially-copyable-corner-cases/).
+
+It's very common to write code that will `memcpy()` a type based on its size. Something like:
+
+```cpp
+S s;
+S s2;
+memcpy(&s2, &s, sizeof(S));
+```
+
+This works great! Until it doesn't! There's one way this could go wrong in older versions of C++,
+and a new way for this to go wrong in C++20.
+
+##### Base classes
+
+The empty base class optimization allows a class type's size to be treated as 0 when it is inherited
+from. Typically C++ does not have zero-sized types. Every object must have a unique address, which
+means it must have a size of at least 1 byte.
+
+```cpp
+struct S {};
+static_assert(sizeof(S) == 1);
+```
+
+But the language relaxes this specifically for a base class. Here the size of `T` is still 1, even
+though the size of `S` is also 1:
+
+```cpp
+struct S {};
+struct T : public S {};
+static_assert(sizeof(T) == 1);
+```
+
+And to be clear that it's the _base_ class which has a zero size, in the following, the size
+of `T` becomes 4, which is the size of its member. There is no extra byte for its base class `S`:
+
+```cpp
+struct S {};
+struct T : public S { int32_t a; };
+static_assert(sizeof(T) == 4);
+```
+
+This presents the first case where the data size of an object matters. If we were to
+`memcpy(&s, &from, sizeof(S))` into an `S*` but it so happens that the object is a `T`, we would
+overwrite one byte of `T::a` with garbage!
+
+The [Clang](https://godbolt.org/z/oejGWP694) and [GCC](https://godbolt.org/z/jjxcMYd3n) (but not
+[MSVC 19](https://godbolt.org/z/7fczM93Gn)) compilers have taken this further, and will generally
+make use of the tail padding in any base class that is not a [Standard-Layout](
+https://en.cppreference.com/w/cpp/language/classes#Standard-layout_class) type.
+
+Recall our earlier example of a struct with tail padding. We mark the `b` member as `private` in
+order to make the type not [Standard-Layout](
+https://en.cppreference.com/w/cpp/language/classes#Standard-layout_class):
+
+```cpp
+struct S {
+    int32_t a;
+  private:
+    int8_t b;
+    // 3 bytes of tail padding.
+};
+static_assert(sizeof(S) == 8);
+```
+
+Then if we make a subclass of `S`, the compiler is entitled to place members into the tail padding
+of `S`, as we see in `T` below:
+
+```cpp
+struct T : S {
+    int8_t c;
+    // 2 bytes of tail padding.
+};
+static_assert(sizeof(S) == 8);
+static_assert(sizeof(T) == 8);
+```
+
+The size of `T` is the same as `S` because the member `c` has been placed inside the tail padding of
+the base class `S`. If we were to `memcpy(&s, &from, sizeof(S))` into an `S*` that has been
+subclassed by `T`, we would overwrite one byte of `T::c` with garbage! However if we

```

Comment:
"an `S*` that is actually a `T`"? Or maybe it'd be clearer to `memcpy(&t, &a, size of(&s))`?

Reply:
Reworded a bit, thanks. I didn't take this verbatim but hopefully it's better.

"into an `S*` that is pointing to a `T` subclass, we would overwrite one byte of `T::c` with garbage, producing a memory safety bug!"